spec/light-client/Blockchain.tla

```diff
@@ -0,0 +1,212 @@
+
+(* A commit is just a set of nodes who have committed the block *)
+Commits == SUBSET AllNodes
+
+(* The set of all block headers that can be on the blockchain.
+   This is a simplified version of the Block data structure in the actual implementation. *)
+BlockHeaders == [
+  height: Heights,
+    \* the block height
+  lastCommit: Commits,
+    \* the nodes who have voted on the previous block, the set itself instead of a hash
+  (* in the implementation, only the hashes of V and NextV are stored in a block,
+     as V and NextV are stored in the application state *) 
+  VP: UNION {Nodes \in SUBSET AllNodes \ {{}}: [Nodes -> Powers]},
+    \* the validators of this block together with their voting powers,
+    \* i.e., a multi-set. We store the validators instead of the hash.
+  NextVP: UNION {Nodes \in SUBSET AllNodes \ {{}}: [Nodes -> Powers]}
+    \* the validators of the next block together with their voting powers,
+    \* i.e., a multi-set. We store the next validators instead of the hash.
+]
+
+(* A signed header is just a header together with a set of commits *)
+\* TODO: Commits is the set of PRECOMMIT messages
+SignedHeaders == BlockHeaders \X Commits
+
+VARIABLES
+    tooManyFaults,
+    (* whether there are more faults in the system than the blockchain can handle *)
+    height,
+    (* the height of the blockchain, starting with 0 *)
+    minTrustedHeight,
+    (* The global height of the oldest block that is younger than
+       the trusted period (AKA the almost rotten block).
+       In the implementation, this is the oldest block,
+       where block.bftTime + trustingPeriod >= globalClock.now. *)
+    blockchain,
+    (* A sequence of BlockHeaders,
+       which gives us the God's (or Daemon's) view of the blockchain. *)
+    Faulty
+    (* A set of faulty nodes, which can act as validators. We assume that the set
+       of faulty processes is non-decreasing. If a process has recovered, it should
+       connect using a different id. *)
+       
+(* all variables, to be used with UNCHANGED *)       
+vars == <<tooManyFaults, height, minTrustedHeight, blockchain, Faulty>>         
+
+(* The set of all correct nodes in a state *)
+Corr == AllNodes \ Faulty       
+
+(****************************** BLOCKCHAIN ************************************)
+(* in the future, we may extract it in a module on its own *)
+
+(* Compute the total voting power of a subset of validators Nodes,
+   whose individual voting powers are given by a function vp *)  
+RECURSIVE PowerOfSet(_, _)
+PowerOfSet(vp, Nodes) ==
+    IF Nodes = {}
+    THEN 0
+    ELSE LET node == CHOOSE n \in Nodes: TRUE IN
+        (* ASSERT(node \in DOMAIN vp) *)
+        vp[node] + PowerOfSet(vp, Nodes \ {node})
+
+TwoThirds(vp, set) ==
+    LET TP == PowerOfSet(vp, DOMAIN vp)
+        SP == PowerOfSet(vp, set)
+    IN
+    3 * SP > 2 * TP 
+
+IsCorrectPowerForSet(Flt, vp, set) ==
+    LET FV == Flt \intersect set
```

Comment:
Can we use full names instead? I am thinking from the perspective of an engineer reading a spec and how we can make the experience better.